packages/passport-interface/src/api/requestCheckIn.ts

```diff
@@ -0,0 +1,72 @@
+import { EdDSATicketPCD, EdDSATicketPCDPackage } from "@pcd/eddsa-ticket-pcd";
+import { ArgumentTypeName } from "@pcd/pcd-types";
+import { SemaphoreIdentityPCDPackage } from "@pcd/semaphore-identity-pcd";
+import { SemaphoreSignaturePCDPackage } from "@pcd/semaphore-signature-pcd";
+import { Identity } from "@semaphore-protocol/identity";
+import urlJoin from "url-join";
+import {
+  CheckTicketInError,
+  CheckTicketInRequest,
+  CheckTicketInResponseValue,
+  ISSUANCE_STRING
+} from "../RequestTypes";
+import { APIResult } from "./apiResult";
+import { httpPost } from "./makeRequest";
+
+/**
+ * Tries to check the user in. This is called by the PCDpass client when
+ * a 'superuser' of a particular event wants to check in a Devconnect
+ * attendee into the event.
+ *
+ * Never rejects. All information encoded in the resolved response.
+ */
+export async function requestCheckIn(
+  passportServerUrl: string,
+  postBody: CheckTicketInRequest
+): Promise<CheckTicketInResult> {
+  return httpPost<CheckTicketInResult>(
+    urlJoin(passportServerUrl, "/issue/check-in"),
+    {
+      // @todo - here and elsewhere - how can we do better than casting, and actually
+      // make sure that the response we're getting back is the right shape?
+      onValue: async (resText) => JSON.parse(resText) as CheckTicketInResult,
+      onError: async () => ({ error: { name: "ServerError" }, success: false })
```

Comment:
Why is this only ever ServerError?  Not NetworkError?  Not AlreadyCheckedIn?
I think I've figured out that server-side errors actually come back as a result which is passed to onValue but is sometimes actually an error and thus doesn't contain a `value` field?  That seems confusing.  `onValue()` actually means any 200 response, which could contain an error?  While `onError()` means a non-200 response or else an exception?  Maybe separate those 3 cases?  Or at least clarify the name of `onValue()` to be `onResult` or `onResponse()` or `on200Response()`?

Reply:
On thinking this through I think the reason this confused me so much is that the construction of an APIResult from an HTTP response is quietly creating two different data types based on the value of the input data (not the type of the input data).  That's an aspect of "structural typing" in TypeScript I haven't experienced before.  The idea of types which are differentiated by data (string) values is unusual.  So it surprised me that a piece of code could create 2 different output types (in a union) from raw data without anything in the code visibly selecting between them.

I think my remaining comment/suggestion here is to clarify terminology.  It's not clear in isolation which of the words "Result", "Value", or "Response" imply something which isn't an error.  Similarly it's not clear in isolation whether an HTTP 400 response is an "Error" or a "Result" or something else.  I don't have a clear suggestion of the best terminology, but it should be consistent, and explained in the comments somewhere.